### PR TITLE
3.2 raft pre voting

### PR DESCRIPTION
**Pull request by andrewkerr9000**
Includes introduction of a visitor for raft messages used by message handling and marshalling (not unmarshalling as that is against the enum value, could be done with another visitor...). The visitor is safer (cannot forget to handle a message, cannot forget last break/else, no casting) and can be used in inheritance, but is a little more verbose and may be counter intuitive to someone not familiar with the pattern.

changelog: Implement a pre-election stage in Raft protocol
This should reduce the incidence of unnecessary elections caused by a disruptive follower, for example rejoining the cluster after a network partition.
This is a breaking change and must be explicitly enabled with `causal_clustering.enable_pre_voting=true`.

### CODE REVIEW

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java [hunk 1]
```diff
@@ -80,62 +78,239 @@ private static void handleLeaderLogCompaction( ReadableRaftState ctx, Outcome ou
...
+    private abstract static class Visitor implements RaftMessages.Visitor<Outcome, IOException>
+    {
+
+        protected final ReadableRaftState ctx;
+        protected final Log log;
+        protected final Outcome outcome;
+
+        public Visitor( ReadableRaftState ctx, Log log )
+        {
+
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java]: superfluous space here and other places

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java [hunk 2]
```diff
@@ -37,6 +37,26 @@
 
 public interface RaftMessages
 {
+
+    interface Visitor<T, E extends Exception>
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java]: Maybe call this a Handler or MessageHandler? And the respective handler methods handle( .. )?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java [hunk 3]
```diff
@@ -59,12 +79,16 @@
...
+
+        PRE_VOTE_REQUEST,
+        PRE_VOTE_RESPONSE,
     }
 
     interface RaftMessage extends Message
     {
         MemberId from();
         Type type();
+        <T, E extends Exception> T visit( Visitor<T, E> visitor ) throws E;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java]: I usually call this "dispatch", as in dispatch message onto correct handler.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Candidate.java [hunk 4]
```diff
@@ -38,124 +38,178 @@
...
-        Outcome outcome = new Outcome( CANDIDATE, ctx );
+        return message.visit( new Visitor( ctx, log ) );
+    }
 
-        switch ( message.type() )
-        {
-            case HEARTBEAT:
-            {
-                RaftMessages.Heartbeat req = (RaftMessages.Heartbeat) message;
+    private static class Visitor implements RaftMessages.Visitor<Outcome, IOException>
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Candidate.java]: I wonder if this static inner class could be merged with the containing class.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Candidate.java [hunk 5]
```diff
@@ -38,124 +38,178 @@
...
             }
+            return outcome;
+        }
 
-            default:
-                break;
+        @Override
+        public Outcome apply( RaftMessages.PreVote.Request request ) throws IOException
+        {
+            return outcome;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Candidate.java]: It's really good that we expose places were we do nothing, so that it doesn't get hidden.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java [hunk 6]
```diff
@@ -47,32 +47,32 @@
...
 
     private MemberId leader;
     private Set<MemberId> votesForMe = new HashSet<>();
+    private Set<MemberId> preVotesForMe = new HashSet<>();
     private Set<MemberId> heartbeatResponses = new HashSet<>();
     private FollowerStates<MemberId> followerStates = new FollowerStates<>();
     private long leaderCommit = -1;
     private long commitIndex = -1;
     private long lastLogIndexBeforeWeBecameLeader = -1;
+    private boolean isPreElection = true; // Needed to form a cluster
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java]: Why is this needed? Is this for unit tests? Perhaps unit tests should generally run with pre-election support off then? ITs etc should be able to form clusters even with the pre-election protocol extension in effect.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/RaftMessageEncoder.java [hunk 7]
```diff
@@ -53,25 +53,67 @@ protected synchronized void encode( ChannelHandlerContext ctx,
         channel.putInt( message.type().ordinal() );
         memberMarshal.marshal( message.from(), channel );
 
-        if ( message instanceof RaftMessages.Vote.Request )
+        message.visit( new Visitor( marshal, memberMarshal, channel ) );
+    }
+
+    private static class Visitor implements RaftMessages.Visitor<Void, Exception>
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/RaftMessageEncoder.java]: Perhaps a better name than just Visitor. EncodingHandler?

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/PreElectionIT.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import org.neo4j.test.causalclustering.ClusterRule;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.not;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+public class PreElectionIT
+{
+    private long electionTimeout = 1;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/PreElectionIT.java]: Maybe just String electionTimeout = "1s"?

**Review by martinfurmanski — COMMENTED** [review]
Looks good overall, but a few minor discussion points.

**Comment by martinfurmanski** [comment]
CI runs indicate that files need to get the correct licenses in place. `mvn license:format` also probably fails `mvn checkstyle:check`

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java (hunk 2 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java]: I'll rename to `Handler`. The outer class name gives context as to what it is handling.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java (hunk 3 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java]: :+1:

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Candidate.java (hunk 4 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Candidate.java]: ReadableRaftState and Log are constructor parameters for the inner visitor class. If it weren't a class they would have to be passed into the dispatch method, which would make the visitor interface less general or rather more complex.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java (hunk 1 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/roles/Follower.java]: :+1:

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java (hunk 6 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java]: This is needed for integration tests. If a cluster member is not in pre-election state when pre-election support is enabled then it won't give a positive pre-vote response to a pre-vote request, and a real election won't occur.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/RaftMessageEncoder.java (hunk 7 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/RaftMessageEncoder.java]: The outer class name spaces the visitor, so it's clear that it refers to raft message encoding. I've renamed it to Handler.

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/PreElectionIT.java (hunk 8 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/PreElectionIT.java]: I was going to reuse it as a long at some point.

I'm still working on these tests.

**Comment by andrewkerr9000** [comment]
Squashed, hopefully ready to merge now.

Integration tests proving troublesome on TC. I've increased the timeout to 5 (!) minutes. One test passes on timeout, so will run for 5 (!) minutes.

**Comment by andrewkerr9000** [comment]
Soak tests all pass reliably

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java (hunk 6 above)
Comment (martinfurmanski) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java]: Hmm.. I think I still don't get why this needs to be set. The pre-election state should not have to be set to give a positive pre-vote response? And an instance seeking an election should be able to reach the isPreElection status quite quickly right?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java [hunk 9]
```diff
@@ -46,32 +46,39 @@
...
         this.voteStorage = voteStorage;
         this.membership = membership;
         this.entryLog = entryLog;
         this.inFlightCache = inFlightCache;
-        log = logProvider.getLog( getClass() );
+        this.supportPreVoting = supportPreVoting;
+        this.log = logProvider.getLog( getClass() );
+
+        // Initial state
+        this.isPreElection = supportPreVoting;
```
Comment (andrewkerr9000) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/state/RaftState.java:81]: Hopefully this is a bit clearer

**Review by martinfurmanski — APPROVED** [review]
